File: judge-api/src/main/kotlin/io/leonfoliveira/judge/api/dto/response/submission/SubmissionPublicResponseDTO.kt
package io.leonfoliveira.judge.api.dto.response.submission

import io.leonfoliveira.judge.api.dto.response.member.MemberPublicResponseDTO
import io.leonfoliveira.judge.api.dto.response.member.toPublicResponseDTO
import io.leonfoliveira.judge.api.dto.response.problem.ProblemPublicResponseDTO
import io.leonfoliveira.judge.api.dto.response.problem.toPublicResponseDTO
import io.leonfoliveira.judge.core.domain.entity.Submission
import io.leonfoliveira.judge.core.domain.enumerate.Language
import java.time.OffsetDateTime
import java.util.UUID

data class SubmissionPublicResponseDTO(
    val id: UUID,
    val member: MemberPublicResponseDTO,
    val problem: ProblemPublicResponseDTO,
    val language: Language,
    val answer: Submission.Answer,
    val createdAt: OffsetDateTime,
)

fun Submission.toPublicResponseDTO(): SubmissionPublicResponseDTO {
    return SubmissionPublicResponseDTO(
        id = id,
        member = member.toPublicResponseDTO(),
        problem = problem.toPublicResponseDTO(),
        language = language,
        answer = answer,
        createdAt = createdAt,
    )
}

File: applications/webapp/src/core/port/driven/repository/dto/response/submission/SubmissionPublicResponseDTO.ts
import { Language } from "@/core/domain/enumerate/Language";
import { SubmissionAnswer } from "@/core/domain/enumerate/SubmissionAnswer";
import { SubmissionStatus } from "@/core/domain/enumerate/SubmissionStatus";
import { MemberPublicResponseDTO } from "@/core/port/driven/repository/dto/response/member/MemberPublicResponseDTO";
import { ProblemPublicResponseDTO } from "@/core/port/driven/repository/dto/response/problem/ProblemPublicResponseDTO";

export type SubmissionPublicResponseDTO = {
  id: string;
  problem: ProblemPublicResponseDTO;
  member: MemberPublicResponseDTO;
  language: Language;
  status: SubmissionStatus;
  answer: SubmissionAnswer;
  createdAt: string;
};
